### PR TITLE
[reporters] Contribute a Slack reporter

### PR DESCRIPTION
**Pull request by artagnon**
The Slack reporter uses BuildRequests to display the build name, and
attempts to use SourceTimestamps to display author name, when they're
available.

Also fix a couple of minor things in the documentation while at it.

Signed-off-by: Ramkumar Ramachandra <artagnon@gmail.com>

### CODE REVIEW

File: master/buildbot/db/sourcestamps.py [hunk 1]
```diff
@@ -38,7 +38,7 @@ class SsList(list):
 
 
 class SourceStampsConnectorComponent(base.DBConnectorComponent):
-    # Documentation is in developer/db.rst
+    # Documentation is in developer/database.rst
```
Comment (cmouse) [inline master/buildbot/db/sourcestamps.py]: should be in it's own commit

File: master/docs/developer/database.rst [hunk 2]
```diff
@@ -210,7 +210,7 @@ builds
 
     Builds are indexed by *buildid* and their contents represented as *builddicts* (build dictionaries), with the following keys:
 
-    * ``id`` (the build ID, globally unique)
+    * ``buildid`` (the build ID, globally unique)
```
Comment (cmouse) [inline master/docs/developer/database.rst]: should be in it's own commit

File: master/buildbot/www/change_hook.py [hunk 3]
```diff
@@ -114,23 +114,26 @@ def getAndSubmitChanges(self, request):
     def makeHandler(self, dialect):
         """create and cache the handler object for this dialect"""
         if dialect not in self.dialects:
-            m = "The dialect specified, '{}', wasn't whitelisted in change_hook".format(dialect)
+            m = "The dialect specified, '{}', wasn't whitelisted in change_hook".format(
```
Comment (cmouse) [inline master/buildbot/www/change_hook.py]: all the changes in change_hook.py should be in it's own commit

**Comment by rodrigc** [comment]
In future, you should do all your changes on a branch and *then* create a pull request.
You should avoid creating a pull request from changes you make directly on the *master* branch.

File: master/buildbot/reporters/slack.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+from datetime import datetime
+from humanize import naturaltime
+from buildbot.status.builder import SUCCESS
+from buildbot.util import httpclientservice
+from buildbot.reporters.http import HttpStatusPushBase
+
+SLACK_BASE_URL = "https://hooks.slack.com"
+
+
+class SlackStatusPush(HttpStatusPushBase):
```
Comment (tardyp) [inline master/buildbot/reporters/slack.py]: This should rather derive from notifierBase, which has more features for text based notifiers

File: master/buildbot/reporters/slack.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        if not isinstance(icon, string_types):
+            config.error('icon must be a string')
+        if not isinstance(notify_on_success, bool):
+            config.error('notify_on_success must be a bool')
+        if not isinstance(notify_on_failure, bool):
+            config.error('notify_on_failure must be a bool')
+
+    @defer.inlineCallbacks
+    def reconfigService(self, endpoint, username="buildbot",
+                        icon="https://buildbot.net/img/nut.png",
```
Comment (tardyp) [inline master/buildbot/reporters/slack.py]: Please don't point to our community infra.

File: master/buildbot/reporters/slack.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            return
+
+        if not self.notify_on_failure and build['result'] != SUCCESS:
+            return
+
+        url = "%s#/builders/%d/builds/%d" % (self.master.config.buildbotURL,
+                                             build['builderid'],
+                                             build['buildid'])
+        naive_startedAt = build['started_at'].replace(tzinfo=None)
+        humantime = naturaltime(datetime.now() - naive_startedAt)
```
Comment (tardyp) [inline master/buildbot/reporters/slack.py]: why don't use complete_at?

File: master/buildbot/reporters/slack.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        ss = yield self.master.db.sourcestamps.getSourceStamp(build['buildid'])
+        br = yield self.master.db.buildrequests.getBuildRequest(build['buildrequestid'])
+
+        if ss:
+            message = """
+Build {name} {result} | {author} | {humantime}
```
Comment (tardyp) [inline master/buildbot/reporters/slack.py]: notifierbase handle the configurable templating with Jinja

File: master/buildbot/reporters/slack.py (hunk 6 above)
Comment (artagnon) [reply master/buildbot/reporters/slack.py]: Wouldn't `complete_at` always give `1 second ago`, since the notification is pushed immediately after completion? We want to know when the request was submitted.

File: master/buildbot/reporters/slack.py (hunk 5 above)
Comment (artagnon) [reply master/buildbot/reporters/slack.py]: It requires a URL, so what do you suggest I point to?

Comment (tardyp) [reply master/buildbot/reporters/slack.py]: I would put a default pointing to the buildbotURL, and make a check to make sure people don't point it to buildbot.net

File: master/buildbot/reporters/slack.py (hunk 6 above)
Comment (tardyp) [reply master/buildbot/reporters/slack.py]: complete_at - started_at!

File: master/buildbot/reporters/slack.py (hunk 4 above)
Comment (artagnon) [reply master/buildbot/reporters/slack.py]: I had the worst development experience ever! Instead of simple NameErrors, I'd get a cryptic

```
	twisted.internet.defer.FirstError: FirstError[#0, [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.defer.FirstError'>: FirstError[#0, [Failure instance: Traceback: <class 'AttributeError'>: 'SlackPush' object has no attribute '_buildsetCompleteConsumer'
```

... and worst of all, `yield`s seem to swallow up even logging messages that are supposed to execute before them (when there's a silly error).

I sank maybe five hours, and endured much frustration in porting the code. I don't know if it's the difference between `HttpStatusPushBase` and `NotifierBase`, or if it's due to the recent Buildbot upgrade, but stuff needs to be fixed.

File: master/buildbot/reporters/slack.py (hunk 5 above)
Comment (artagnon) [reply master/buildbot/reporters/slack.py]: I can't find nut.png anywhere in buildbotURL. Can you point me to it?

Comment (tardyp) [reply master/buildbot/reporters/slack.py]: there is $buildbotURL/img/favicon.ico

Comment (tardyp) [reply master/buildbot/reporters/slack.py]: but we can add a larger version by creating it in www/base/src/img/
